File: solutions/cpp/valid-parentheses.cc
// Copyright (c) 2013 Elements of Programming Interviews. All rights reserved.

#include <cassert>
#include <iostream>
#include <stack>
#include <string>

using std::boolalpha;
using std::cout;
using std::endl;
using std::stack;
using std::string;

// @include
bool IsValidParentheses(const string& s) {
  stack<char> left_parens;
  for (size_t i = 0; i < s.size(); ++i) {
    if (s[i] == '(' || s[i] == '{' || s[i] == '[') {
      left_parens.emplace(s[i]);
    } else {
      if (left_parens.empty()) {
        return false;  // No left paren.
      }
      if ((s[i] == ')' && left_parens.top() != '(') ||
          (s[i] == '}' && left_parens.top() != '{') ||
          (s[i] == ']' && left_parens.top() != '[')) {
        return false;  // Unmatched left paren.
      }
      left_parens.pop();
    }
  }
  return left_parens.empty();
}
// @exclude

void small_test() {
  assert(IsValidParentheses("()"));
  assert(IsValidParentheses("()[]{}"));
  assert(IsValidParentheses("[()[]]{}"));
  assert(IsValidParentheses("(()[]{()[]{}{}})"));
  assert(!IsValidParentheses("([)]"));
  assert(!IsValidParentheses("["));
  assert(!IsValidParentheses("(()[]{()[]{})({}})"));
}

int main(int argc, char** argv) {
  small_test();
  if (argc == 2) {
    cout << boolalpha << IsValidParentheses(argv[1]) << endl;
  }
  return 0;
}

File: solutions/python/valid_parentheses.py
import sys


# @include
def is_well_formed(s):
    left_chars, lookup = [], {'(': ')', '{': '}', '[': ']'}
    for c in s:
        if c in lookup:
            left_chars.append(c)
        elif not left_chars or lookup[left_chars.pop()] != c:
            # Unmatched right char or mismatched chars.
            return False
    return not left_chars
# @exclude


def small_test():
    assert is_well_formed('()')
    assert is_well_formed('()[]{}')
    assert is_well_formed('[()[]]{}')
    assert is_well_formed('(()[]{()[]{}{}})')
    assert not is_well_formed('([)]')
    assert not is_well_formed('[')
    assert not is_well_formed('(()[]{()[]{})({}})')


def main():
    small_test()
    if len(sys.argv) == 2:
        print(is_well_formed(sys.argv[1]))


if __name__ == '__main__':
    main()
